palantir-java-format-native/build.gradle

```diff
@@ -0,0 +1,111 @@
+    }
+}
+
+
+apply plugin: 'java-library'
+apply plugin: 'org.graalvm.buildtools.native'
+apply plugin: 'com.palantir.external-publish-custom'
+
+dependencies {
+    implementation project(':palantir-java-format')
+}
+
+import com.palantir.platform.OperatingSystem
+import com.palantir.platform.Architecture
+
+def operatingSystem = OperatingSystem.get()
+def architecture = Architecture.get()
+
+def exports = [
+        'jdk.compiler/com.sun.tools.javac.file',
+        'jdk.compiler/com.sun.tools.javac.main',
+        'jdk.compiler/com.sun.tools.javac.parser',
+        'jdk.compiler/com.sun.tools.javac.tree',
+        'jdk.compiler/com.sun.tools.javac.util',
+        'jdk.compiler/com.sun.tools.javac.code',
+        'jdk.compiler/com.sun.tools.javac.api'
+]
+
+def jvmArgList = exports.collect { value -> "--add-exports=${value}=ALL-UNNAMED".toString() }
+
+
+graalvmNative {
+    binaries {
+        main {
+            javaLauncher = javaToolchains.launcherFor {
+                languageVersion = JavaLanguageVersion.of(23)
+                vendor = JvmVendorSpec.GRAAL_VM
+            }
+            imageName = String.format('palantir-java-format-%s-%s', operatingSystem.uiName(), architecture.uiName())
+            mainClass = 'com.palantir.javaformat.java.Main'
+            fallback = false
+            sharedLibrary = false
+            classpath.setFrom(configurations.named('runtimeClasspath'))
+            configurationFileDirectories.from(file('src/main/resources/META-INF/native-image'))
+            buildArgs.addAll("-H:ConfigurationFileDirectories=${projectDir}/src/main/resources/META-INF/native-image",
+                    '-H:+UnlockExperimentalVMOptions',
+                    '-H:IncludeResourceBundles=com.sun.tools.javac.resources.compiler',
+                    '-H:IncludeResourceBundles=com.sun.tools.javac.resources.javac',
+                    '--no-fallback',
+                    '--initialize-at-build-time=com.sun.tools.javac.file.Locations',
+                    '-H:+ReportExceptionStackTraces',
+                    '-H:-UseContainerSupport',
+                    "-march=compatibility")
+            jvmArgs.addAll(jvmArgList)
+        }
+    }
+}
+
+
+// The image requires parameters to be passed to; eg. gw nativeRun -Pargs=--palantir,-i,<file>
+tasks.named("nativeRun", org.graalvm.buildtools.gradle.tasks.NativeRunTask) {
+    runtimeArgs.addAll(project.hasProperty('appArgs') ? project.appArgs.split(',') : [])
+}
+
+def osExtension = operatingSystem == OperatingSystem.MACOS ? 'bin' : 'exe'
+
+externalPublishing {
+    publishing {
+        publications {
+            nativeImage(MavenPublication) {
+                fileTree("build/native/nativeCompile", { it.include("**/*") }).forEach { file ->
+                    artifact(file) {
+                        classifier = String.format("nativeImage-%s", getOsArchName(file))
+                        extension = osExtension
+                    }
+                }
```

Comment:
This feels odd with it not being lazy. If the tasks have not run yet when this is evaluated then there will be no artifacts for this publication. I guess it works because we attach workspaces before gradle is run for publish. But does this mean `publishToMavenLocal` run on a laptop doesn't work?

Reply:
`artifact` can apparently take anything that `Project#file` can, so we can give it providers or other tasks - I'm just not sure if it will try to rerun the tasks or not on the publish task (which won't work). If publishToMavenLocal doesn't currently work, maybe we need a separate publication not registered with external-publish-plugin that actually takes the tasks as inputs to the artifacts?

Reply:
It works locally once I added: https://github.com/palantir/palantir-java-format/pull/1208/commits/8456625f22b51b7930eb83adf6fb2ea640f5a5ab#diff-012bca426d515b4e67fdc32e660cf6ade9abd1d720f3f923e6b16101a72dc3c3R84

Reply:
ended up doing 2 separate publication depending if we are in CI or not - https://github.com/palantir/palantir-java-format/pull/1208/files#diff-012bca426d515b4e67fdc32e660cf6ade9abd1d720f3f923e6b16101a72dc3c3R77